# Проект: "Исследование надежности заемщиков"


В проекте выполнена оценка надежности заемщиков по данным, предоставленным банком.

В качестве характеристик заемщиков, которые могут оказывать влияние на возврат кредита в срок, исследованы:

- количество детей у заемщика; 
- его семейное положение;
- уровень ежемесячного дохода. 

Кроме того, проанализировано влияние цели кредита на его своевременный возврат. 

Полученные результаты предполагается использовать для построения модели кредитного скоринга, которая позволит оценить способность заёмщика вернуть кредит.

**Расшифровка значения столбцов в датасете от банка:**

`children` — количество детей в семье

`days_employed` — общий трудовой стаж в днях

`dob_years` — возраст клиента в годах

`education` — уровень образования клиента

`education_id` — идентификатор уровня образования

`family_status` — семейное положение

`family_status_id` — идентификатор семейного положения

`gender` — пол клиента

`income_type` — тип занятости

`debt` — имел ли задолженность по возврату кредитов (1 - да, 0 - нет)

`total_income` — ежемесячный доход

`purpose` — цель получения кредита

**План выполнения проекта:**

1. Выполнить загрузку датасета из csv-файла и первичный просмотр информации о датасете.
2. Выполнить предобработку данных: удалить пропуски, дубликаты, выполнить категоризацию заемщиков по уровню дохода.
3. Проанализировать влияние индивидуальных характеристик заемщика и кредитной цели на возврат кредита в срок.
4. Сделать общие выводы о влиянии изученных характеристик заемщиков на своевременный возврат кредита.

## Первичный просмотр предоставленных данных

**Импортируем библиотеку pandas и считаем данные из csv-файла в датафрейм, сохранив его в переменную `data`:**

In [1]:
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

data = pd.read_csv('data.csv')

**Выведем на экран первые 20 строчек датафрейма `data`.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Получим основную информацию о датафрейме:**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Оценим количество пропущенных значений для каждого столбца:**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**В двух столбцах есть пропущенные значения - `total_income` и `days_employed`. Для начала детально рассмотрим данные о доходе из непустых ячеек `total income`:**

In [5]:
data['total_income'].describe()

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64

**Данные о доходе из непустых ячеек выглядят правдоподобно: минимум 20667, максимум 2265604, медиана 142594, среднее 165343. Поскольку на сумму дохода сильнее всего влияет тип занятости, заполним пропуски в столбце `total_income` медианным значением по каждому типу занятости из столбца `income_type`.**

In [6]:
for income_type in data['income_type'].unique():
    data.loc[(data['income_type'] == income_type) & (data['total_income'].isna()), 'total_income'] \
    = data.loc[data['income_type'] == income_type, 'total_income'].median()

**Проверим, что пропуски в `total_income` успешно заполнены:**

In [7]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

**Заменим вещественный тип данных в столбце `total_income` на целочисленный**

In [8]:
data['total_income'] = data['total_income'].astype(int)
data['total_income'].dtype

dtype('int32')

**Проверим, как распределены данные в столбце трудового стажа -  `days_employed`**

In [9]:
data['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

**С данными о трудовом стаже точно есть проблемы: во-первых, присутствуют отрицательные значения стажа, во-вторых, есть неоправданно завышенне цифры стажа. Поскольку трудовой стаж не является интересующим нас параметром, влияющим на возврат кредита в срок, удалим столбец `days_employed` из датасета.**

In [10]:
data = data.drop('days_employed', axis=1)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   dob_years         21525 non-null  int64 
 2   education         21525 non-null  object
 3   education_id      21525 non-null  int64 
 4   family_status     21525 non-null  object
 5   family_status_id  21525 non-null  int64 
 6   gender            21525 non-null  object
 7   income_type       21525 non-null  object
 8   debt              21525 non-null  int64 
 9   total_income      21525 non-null  int32 
 10  purpose           21525 non-null  object
dtypes: int32(1), int64(5), object(5)
memory usage: 1.7+ MB


**Выведем перечень уникальных значений столбца `children`.**

In [11]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие значения, из датафрейма `data`.**

In [12]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [13]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Обработка дубликатов

**Обработаем неявные дубликаты в столбце `education` - в нем есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру:**

In [14]:
data['education'] = data['education'].str.lower()
data['education'].head(10)

0     высшее
1    среднее
2    среднее
3    среднее
4    среднее
5     высшее
6     высшее
7    среднее
8     высшее
9    среднее
Name: education, dtype: object

**Проверим, есть ли в датасете строки, полностью дублирующие друг-друга:**

In [15]:
data.duplicated().sum()

71

**Удалим такие строки:**

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Создадим в датафрейме `data` новый столбец, в котором укажем категорию дохода заемщика. Будем руководствоваться следующими диапазонами:**

- 0–30000 — категория `'E'`;
- 30001–50000 — категория `'D'`;
- 50001–200000 — категория`'C'`;
- 200001–1000000 — категория `'B'`;
- 1000001 и выше — категория `'A'`.

**Для этого создадим функцию `categorize_income`, которая в зависимости от передаваемого ей значения, будет возвращать обозначение соответствующей категории дохода:** 

In [17]:
def categorize_income(income):
    if income <= 30000:
        return 'E'
    if income <= 50000:
        return 'D'
    if income <= 200000:
        return 'C'
    if income <=1000000:
        return 'B'
    return 'A'

**Применим функцию `categorize_income` к столбцу `total_income` датафрейма `data`:**

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Посмотрим появился ли в `data` новый столбец:**  

In [19]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21331 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   children               21331 non-null  int64 
 1   dob_years              21331 non-null  int64 
 2   education              21331 non-null  object
 3   education_id           21331 non-null  int64 
 4   family_status          21331 non-null  object
 5   family_status_id       21331 non-null  int64 
 6   gender                 21331 non-null  object
 7   income_type            21331 non-null  object
 8   debt                   21331 non-null  int64 
 9   total_income           21331 non-null  int32 
 10  purpose                21331 non-null  object
 11  total_income_category  21331 non-null  object
dtypes: int32(1), int64(5), object(6)
memory usage: 2.0+ MB


**Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Распределим цели из столбца `purpose` по четрыем категориям:** 
- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Для этого создадим функцию `categorize_purpose`, которая в зависимости от содержимого в столбце `purpose` будет возвращать одну из обозначенных выше категорий:**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

**Применим функцию `categorize_purpose` к столбцу `purpose`. Результат запишем в новый столбец `purpose_category`:**

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

**Проверим, какие уникальные значения появились в новом столбце**:

In [23]:
data['purpose_category'].unique()

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

## Анализ влияния индивидуальных характеристик заемщика и кредитной цели на возврат кредита в срок

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

**Оценим, в каком диапазоне находятся значения из столбца `children`, и каково распределение клиентов банка по количеству детей. Сохраним полученный результат в Series `number_by_children` и выведем на экран:**

In [24]:
number_by_children = data['children'].value_counts()
number_by_children

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

**Посчитаем количество тех, кто имел задолженность по возврату кредитов, для групп клиентов с разным количеством детей и запишем результат в `debtors_by_children`:**

In [25]:
debtors_by_children = data.loc[data['debt'] == 1].groupby('children')['children'].count()
debtors_by_children

children
0    1063
1     444
2     194
3      27
4       4
Name: children, dtype: int64

**Судя по полученному результату, люди с 5-ю детьми не имели задолженностей по кредитам. Проверим это, выведем таблицу cо списком таких людей:**

In [26]:
print(data.loc[data['children'] == 5, ['children','debt']])

       children  debt
3979          5     0
4397          5     0
7866          5     0
15822         5     0
15916         5     0
16211         5     0
20452         5     0
20837         5     0
21156         5     0


**Добавим в `debtors_by_children` новую строку, в которой отразим, что среди людей с пятью детьми нет тех, у кого задолженность по кредитам:**

In [27]:
debtors_by_children[5] = 0
debtors_by_children

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: children, dtype: int64

**Рассчитаем `debtors_by_children_share` - долю должников в каждой из групп:**

In [28]:
debtors_by_children_share = debtors_by_children/number_by_children
print(round(debtors_by_children_share, 3))

children
0    0.075
1    0.092
2    0.095
3    0.082
4    0.098
5    0.000
Name: children, dtype: float64


**Рассчитаем вероятность возврата кредита в срок - для каждой из групп вычтем из единицы долю должников:**

In [29]:
probability = 1 - debtors_by_children_share
print(round(probability, 3))

children
0    0.925
1    0.908
2    0.905
3    0.918
4    0.902
5    1.000
Name: children, dtype: float64


**Вывод:**
Зависимость между количеством детей и возвратом кредита в срок существует.
- С наибольшей вероятностью (92.5 %) кредит в срок возвращают те, у кого нет детей.
- Для людей у которых от 1 до 4-х детей, вероятность возврата кредита в срок составляет от 90.2 до 91.8 %. Чтобы узнать, различаются ли достоверно эти проценты, нужно применить биномиальный критерий - попарно сравнить с помощью него вероятности в группах.
- Несмотря на то, что согласно нашим расчетам 100 % людей у которых пятеро детей, отдают кредит в срок, этот результат нельзя признать показательным. Выборка из 9 человек слишком мала, для того чтобы делать по ней выводы.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

**Для удобства работы с данными выполним их декомпозицию. В датасет `family_status_id_debt` запишем данные из столбцов `family_status_id` и `debt`, другой датасет `family_status_dict` будет словарем - в нем оставим только столбцы `family_status_id` и `family_status`, и уберем в нем дублирующиеся cтроки:**

**Создадим датасет `family_status_id_debt` c данными из столбцов `family_status_id` и `debt`. Выведем на экран его первые 5 строк:**

In [30]:
family_status_id_debt = data[['family_status_id','debt']]
family_status_id_debt.head() 

family_status_id  debt
0                 0     0
1                 0     0
2                 0     0
3                 0     0
4                 1     0

**Создадим словарь `family_status_dict` со столбцами `family_status_id` и `family_status`. Уберем дубликаты и выведем датасет на экран:**

In [31]:
family_status_dict = data[['family_status_id','family_status']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

**Далее рассчитаем число клиентов с разными `family_status_id` и запишем его в объект Series с названием `number_by_family_status_id`:**

**Отсортируем датасет `family_status_id_debt` по возрастанию `family_status_id` и запишем в новый датасет `family_status_id_debt_sorted`**

In [32]:
family_status_id_debt_sorted = family_status_id_debt.sort_values(by='family_status_id', ascending=True)

**Cгруппируем строки в `family_status_id_debt_sorted` по `family_status_id` и посчитаем сколько клиентов в группах с разным
семейным положением:**

In [33]:
number_by_family_status_id = family_status_id_debt_sorted.groupby('family_status_id').count()
number_by_family_status_id

debt
family_status_id       
0                 12261
1                  4134
2                   951
3                  1189
4                  2796

**Рассчитаем число тех, кто возвращал кредит в срок в каждой из групп и запишем в объект Series `in_time_payers_by_family_status_id`. Для этого сначала оставим в датесете `family_status_id_debt_sorted` только тех, у кого в столбце debt стоит 0:**

In [34]:
in_time_payers = family_status_id_debt_sorted.loc[family_status_id_debt_sorted['debt'] == 0]

**Посчитаем сколько добросовестных заемщиков в каждой группе:**

In [35]:
in_time_payers_by_family_status_id = in_time_payers.groupby('family_status_id').count()
in_time_payers_by_family_status_id

debt
family_status_id       
0                 11334
1                  3749
2                   888
3                  1105
4                  2523

**Вычислим долю добросовестных заемщиков с разным семейным положением:**

In [36]:
in_time_payers_by_family_status_id_share = in_time_payers_by_family_status_id/number_by_family_status_id
print(round(in_time_payers_by_family_status_id_share, 3)) 

                   debt
family_status_id       
0                 0.924
1                 0.907
2                 0.934
3                 0.929
4                 0.902


**Для удобства восприятия информации присоединим к полученной таблице данные о `family_status` из словаря `family_status_dict`:**

In [37]:
in_time_payers_by_family_status_share = \
in_time_payers_by_family_status_id_share.merge(family_status_dict, on ='family_status_id', how ='left')

in_time_payers_by_family_status_share

family_status_id      debt          family_status
0                 0  0.924394        женат / замужем
1                 1  0.906870       гражданский брак
2                 2  0.933754         вдовец / вдова
3                 3  0.929352              в разводе
4                 4  0.902361  Не женат / не замужем

**Удалим колонку `family_status_id`:** 

In [38]:
in_time_payers_by_family_status_share = in_time_payers_by_family_status_share[['debt','family_status']]
in_time_payers_by_family_status_share

debt          family_status
0  0.924394        женат / замужем
1  0.906870       гражданский брак
2  0.933754         вдовец / вдова
3  0.929352              в разводе
4  0.902361  Не женат / не замужем

**Переведем значения в колонке `debt` в проценты и округлим до одного знака после запятой:**

In [39]:
in_time_payers_by_family_status_share['debt'] = round(in_time_payers_by_family_status_share['debt']*100, 1)
in_time_payers_by_family_status_share

debt          family_status
0  92.4        женат / замужем
1  90.7       гражданский брак
2  93.4         вдовец / вдова
3  92.9              в разводе
4  90.2  Не женат / не замужем

**Переименуем колонку `debt` в `in_time_payers_percentage`:**

In [40]:
new_columns_names = {'debt':'in_time_payers_percentage', 'family_status':'family_status'}
in_time_payers_by_family_status_share = in_time_payers_by_family_status_share.rename(columns = new_columns_names)
in_time_payers_by_family_status_share

in_time_payers_percentage          family_status
0                       92.4        женат / замужем
1                       90.7       гражданский брак
2                       93.4         вдовец / вдова
3                       92.9              в разводе
4                       90.2  Не женат / не замужем

**Вывод:**
- Зависимость между семейным положением и возвратом кредита в срок есть. Чаще других (93.4 %) кредиты в срок возвращают вдовы/вдовцы, немного реже - люди, находящиеся в разводе (92.9 %), и замужние/женатые (92.4 %). Самые низкие вероятности возврата кредита в срок демонстрируют люди, состоящие в гражданском браке (90.7 %) и холостяки (90.2 %). 

- Существует ли достоверная разница между этими процентами - вопрос, на который можно ответить, применив стат. методы. анализа. Размеры выборок для этого вполне достаточны. 

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

**Чтобы не работать с большой таблицей, создадим новую, в которой будут только данные о наличии задолженности по кредиту `debt` и категории дохода `total_income_category`:**

In [41]:
income_category_debt = data[['total_income_category', 'debt']].sort_values(by='total_income_category', ascending = True)
income_category_debt.head(10)

total_income_category  debt
4415                      A     1
11071                     A     0
15268                     A     0
18766                     A     0
4129                      A     0
18368                     A     0
1590                      A     0
10004                     A     0
19338                     A     0
13090                     A     0

**Посчитаем число заемщиков в каждой категории дохода:**

In [42]:
number_by_income_category = income_category_debt.groupby('total_income_category').count()
number_by_income_category

debt
total_income_category       
A                         25
B                       5014
C                      15921
D                        349
E                         22

**Найдем число добросовестных заемщиков в каждой категории. Для этого сначала создадим таблицу, в которой присутствуют только те, кто отдал кредит в срок, а затем рассчитаем число добросовестных заемщиков по каждой категории дохода:**

In [43]:
in_time_payers_by_income_category = income_category_debt.loc[income_category_debt['debt'] == 0]
in_time_payers_by_income_category = in_time_payers_by_income_category.groupby('total_income_category').count()
in_time_payers_by_income_category

debt
total_income_category       
A                         23
B                       4660
C                      14568
D                        328
E                         20

**Рассчитаем `in_time_payers_share_by_income_category` - долю добросовестных заемщиков  в каждой категории дохода:**

In [44]:
in_time_payers_share_by_income_category = in_time_payers_by_income_category/number_by_income_category
in_time_payers_share_by_income_category

debt
total_income_category          
A                      0.920000
B                      0.929398
C                      0.915018
D                      0.939828
E                      0.909091

**Переведем значения в колонке `debt` в проценты и округлим до одного знака после запятой:**

In [45]:
in_time_payers_share_by_income_category['debt'] = round(in_time_payers_share_by_income_category['debt']*100, 1)
in_time_payers_share_by_income_category

debt
total_income_category      
A                      92.0
B                      92.9
C                      91.5
D                      94.0
E                      90.9

**Переименуем колонку `debt` в `in_time_payers_percentage`**

In [46]:
new_columns_names = {'total_income_category':'total_income_category', 'debt':'in_time_payers_percentage'}
in_time_payers_share_by_income_category = in_time_payers_share_by_income_category.rename(columns = new_columns_names)
in_time_payers_share_by_income_category

in_time_payers_percentage
total_income_category                           
A                                           92.0
B                                           92.9
C                                           91.5
D                                           94.0
E                                           90.9

**Чтобы было удобнее ориентироваться в категориях дохода, создадим столбец `income_category_description` с расшифровкой:**

In [47]:
in_time_payers_share_by_income_category['income_category_description'] = ['1000001 и выше', '200001-1000000',\
                                                                          '50001-200000', '30001-50000', '0-30000'] 
in_time_payers_share_by_income_category

in_time_payers_percentage income_category_description
total_income_category                                                       
A                                           92.0              1000001 и выше
B                                           92.9              200001-1000000
C                                           91.5                50001-200000
D                                           94.0                 30001-50000
E                                           90.9                     0-30000

**Вывод:**
- Зависимость между уровнем дохода и возвратом кредита в срок есть. Чаще других кредиты в срок возвращают люди с доходом в категории D (30001-50000) - в 94.0 % случаев. Более низкие проценты возврата кредита в срок у клиентов с доходом категорий B (200001-1000000) и С (50001-200000) - 92.9 и 91.5 %, соответственно.
- Проценты возвратов кредита в срок в группах с доходом уровня A и E сложно назвать репрезентативными, поскольку число таких клиентов невелико  - 25 и 22 человека, соответственно. Таким образом, вклад любого из клиентов в этих группах в среднегрупповой процент отдачи кредита в срок составляет 4 и 4.5 % соответственно, что больше выявленных нами различий между группами.   

### Как разные цели кредита влияют на его возврат в срок?

**Выберем из датасета data, только те столбцы, которые содержат информацию, нужную для ответа на поставленный вопрос:**

In [48]:
purpose_debt = data[['purpose_category','debt']]
purpose_debt.head(10)

purpose_category  debt
0  операции с недвижимостью     0
1    операции с автомобилем     0
2  операции с недвижимостью     0
3     получение образования     0
4        проведение свадьбы     0
5  операции с недвижимостью     0
6  операции с недвижимостью     0
7     получение образования     0
8        проведение свадьбы     0
9  операции с недвижимостью     0

**Получим список уникальных целей на получение кредита - `purpose_list`:**

In [49]:
purpose_list = purpose_debt['purpose_category'].unique()
purpose_list

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

**Создадим функцию `purpose_id()`, которая возвращает идентификатор от 1 до 4 в зависимости от категории цели кредита:**

In [50]:
def purpose_id(purpose_category):
    if purpose_category == 'операции с недвижимостью':
        return 1
    elif purpose_category == 'операции с автомобилем':
        return 2
    elif purpose_category == 'получение образования':
        return 3
    elif purpose_category == 'проведение свадьбы':
        return 4
    return 'other'

**При помощи функции `purpose_id()` создадим в датасете `purpose_debt` новый столбец `purpose_category_id`:**

In [51]:
purpose_debt['purpose_category_id'] = purpose_debt['purpose_category'].apply(purpose_id)
purpose_debt

purpose_category  debt  purpose_category_id
0      операции с недвижимостью     0                    1
1        операции с автомобилем     0                    2
2      операции с недвижимостью     0                    1
3         получение образования     0                    3
4            проведение свадьбы     0                    4
...                         ...   ...                  ...
21520  операции с недвижимостью     0                    1
21521    операции с автомобилем     0                    2
21522  операции с недвижимостью     1                    1
21523    операции с автомобилем     1                    2
21524    операции с автомобилем     0                    2

[21331 rows x 3 columns]

**Создадим словарь `purpose_dict`, в котором будут записаны категории целей и их идентификаторы:**

In [52]:
purpose_dict = purpose_debt[['purpose_category_id', 'purpose_category']]

# Удалим дубликаты:
purpose_dict = purpose_dict.drop_duplicates()
purpose_dict

purpose_category_id          purpose_category
0                    1  операции с недвижимостью
1                    2    операции с автомобилем
3                    3     получение образования
4                    4        проведение свадьбы

**Уберем из таблицы `purpose_debt` текстовые описания категорий `purpose_category`:**

In [53]:
purpose_debt = purpose_debt[['purpose_category_id','debt']]
purpose_debt

purpose_category_id  debt
0                        1     0
1                        2     0
2                        1     0
3                        3     0
4                        4     0
...                    ...   ...
21520                    1     0
21521                    2     0
21522                    1     1
21523                    2     1
21524                    2     0

[21331 rows x 2 columns]

**Посчитаем количество заемщиков в каждой категории целей кредитов и запишем в `purpose_number`:**

In [54]:
purpose_number = purpose_debt.groupby('purpose_category_id').count()
purpose_number

debt
purpose_category_id       
1                    10751
2                     4279
3                     3988
4                     2313

**Найдем всех заемщиков, вернувших кредит в срок:**

In [55]:
purpose_in_time = purpose_debt.loc[purpose_debt['debt'] == 0]
purpose_in_time

purpose_category_id  debt
0                        1     0
1                        2     0
2                        1     0
3                        3     0
4                        4     0
...                    ...   ...
21518                    2     0
21519                    1     0
21520                    1     0
21521                    2     0
21524                    2     0

[19599 rows x 2 columns]

**Посчитаем количество добросовестных заемщиков для каждой категории кредита - `purpose_in_time`:**

In [56]:
purpose_in_time = purpose_in_time.groupby('purpose_category_id').count()
purpose_in_time

debt
purpose_category_id      
1                    9971
2                    3879
3                    3619
4                    2130

**Вычислим долю заемщиков, вернувших кредит в срок, по каждой категории кредита:**

In [57]:
purpose_in_time_share = purpose_in_time/purpose_number
purpose_in_time_share

debt
purpose_category_id          
1                    0.927449
2                    0.906520
3                    0.907472
4                    0.920882

**Для наглядности добавим столбец с названиями категорий кредита `purpose_category` в датафрейм `purpose_in_time_share`**

In [58]:
purpose_in_time_share = purpose_in_time_share.merge(purpose_dict, on = 'purpose_category_id', how = 'left')
purpose_in_time_share

purpose_category_id      debt          purpose_category
0                    1  0.927449  операции с недвижимостью
1                    2  0.906520    операции с автомобилем
2                    3  0.907472     получение образования
3                    4  0.920882        проведение свадьбы

**Переведем значения в столбце `debt` в проценты с точностью до одного знака после запятой:**

In [59]:
purpose_in_time_share['debt'] = round(purpose_in_time_share['debt']*100, 1)
purpose_in_time_share

purpose_category_id  debt          purpose_category
0                    1  92.7  операции с недвижимостью
1                    2  90.7    операции с автомобилем
2                    3  90.7     получение образования
3                    4  92.1        проведение свадьбы

**Переименуем столбец `debt` в `in_time_payers_percentage`:**

In [60]:
new_columns_names = {'purpose_category_id':'purpose_category_id', 'debt':'in_time_payers_percentage', \
                     'purpose_category':'purpose_category'}
purpose_in_time_share = purpose_in_time_share.rename(columns = new_columns_names)
purpose_in_time_share

purpose_category_id  in_time_payers_percentage          purpose_category
0                    1                       92.7  операции с недвижимостью
1                    2                       90.7    операции с автомобилем
2                    3                       90.7     получение образования
3                    4                       92.1        проведение свадьбы

**Вывод.**
Цель кредита оказывает влияние на вероятность того, что кредит будет возвращен в срок. Чаще других в срок возвращают кредиты, взятые для выполнения операций с недвижимостью (92.7 %) и проведения свадьбы (92.1 %). Кредиты, взятые для покупки авто и для получения образования возвращают вовремя несколько реже (в 90.7 % случаев).

## Общий вывод.

**`Целью`** настоящего исследования была оценка влияния ряда характеристик потенциальных заемщиков, а также цели, на которую берут кредит, на его погашение в срок. Полученные результаты предполагается использовать для построения модели кредитного скоринга, которая позволит оценить способность заёмщика вернуть кредит.

- **В ходе проекта с привлечением данных о заемщиках, предостаставленной банком, были даны ответы на следующие вопросы:**

1. Как влияет количество детей у заемщика на возврат кредита в срок?
2. Существует ли зависимость между семейным положением заемщика и возвратом кредита в срок?
3. Влияет ли уровень дохода заемщика на своевременный возврат кредита?
4. Как разные кредитные цели влияют на возврат кредита в срок?

- **Для ответа на эти вопросы, на первом этапе исследования была выполнена предобработка данных**:

1. Недостающие данные о доходе и о трудовом стаже индивидуальных заемщиков были заполнены соответствующими медианными значениями для группы заемщиков с таким же типом занятости.
2. Поскольку в данных о количестве детей заемщиков, были обнаружены аномальные значения (-1, 20), информация о таких заемщиках была полностью убрана из датасета.
3. Данные о доходе представлены в виде целых чисел вместо чисел с целой и дробной частью.
4. Добавлены столбцы с категорией дохода заемщика (A, B, C, D, E) и категорией кредита (операции с автомобилем, операции с недвижимостью, проведение свадьбы, получение образования).
5. Унифицированы обозначения уровня образования заемщиков.

_Примечание: В предоставленном банком датасете обнаружены аномальные данные о трудовом стаже для двух групп заемщиков - пенсионеров и студентов. Поскольку эти данные не использовались для выполнения настоящего исследования, с ними не выполняли никаких манипуляций_


**Результаты исследования:**

**`1. Количество детей у заемщика`**

| Количество детей              | **1**     | **2** | **3** |**4** |**Нет детей**| 
| :-----------------------------|:---------:|:-----:|:-----:|:----:|:-----------:|
| **Возврат кредита в срок, %** | 90.8      | 90.5  | 91.8  | 90.2 | **92.5**    |

**Заемщики, у которых нет детей, более надежны, чем те, у кого есть дети. Общая тенденция такова, что с каждым ребенком вероятность возврата кредита снижается на 0.2 п.п.**


**`2. Семейное положение заемщика`**

|**Семейное положение**          |**гражданский брак**|**не женат/не замужем**|**женат/замужем**|**в разводе**|**вдовец/вдова**|
| :------------------------------| :-----------------:|:---------------------:|:---------------:|:-----------:|:-----------:|
| **Возврат кредита в срок, %**  | 90.7               | 90.2                  | **92.4**        | **92.9**    | **93.4**    |

**Заемщики, которые ранее состояли или состоят в браке надежнее чем те, кто не женат/не замужем**


**`3. Доход заемщика`**

В предоставленных банком данных самыми многочисленными группами клиентов были те, чей доход составляет от 500001 до 200000 и от 200001 до 1000000 в месяц. Таких людей было 74 и 24 % от числа всех заемщиков - суммарно 98 %. Группы с доходом 500000 и ниже, и 1000001 и выше составляли только 2 % от общего числа заемщиков. Столь малые выборки не репрезентативны, поэтому обратим внимание только на самые широко представленные группы заемщиков по доходу:   

| Ежемесячный доход             |**50001-200000** |**200001-1000000**| 
| :-----------------------------|:---------------:|:----------------:|
| **Возврат кредита в срок, %** | 91.5            | **92.9**         |

**Заемщики с высоким доходом склонны более добросовестно выполнять кредитные обязательства.**


**`4. Цель кредита`**

| Цель кредита                  | **образование** |**автомобиль** | **свадьба** | **недвижимость** |
| :-----------------------------| :--------------:|:-------------:|:-----------:|:----------------:|
| **Возврат кредита в срок, %** | 90.7            | 90.7          | 92.1        | **92.7**         |

**Чаще других в срок отдают кредиты на недвижимость**

<br>

**Идеальным заемщиком по исследованным нами параметрам можно назвать человека, который состоит или ранее состоял в браке, берущего кредит на покупку недвижимости, с высоким уровнем дохода и без детей. Напротив, наименее надежным следует признать заемщика с несколькими детьми, который не состоит и ранее не состоял в браке, с низким уровнем дохода, берущего автокредит или кредиит на образование. Описанные типажи заемщиков займут крайние положения на шкале кредитного скоринга.**
